### PR TITLE
Some tiny modification about torch.set/get_default_device

### LINKED ISSUE #115333: `torch.set_default_device` does not affect `torch._C._get_default_device`
### 🐛 Describe the bug

See title. With `torch.set_default_tensor_type`, the default device is switched normally. However, that function is deprecated now and its replacement `torch.set_default_device` doesn't affect `torch._C._get_default_device`.

```python3
>>> import torch
torch.>>> torch.__version__
'2.1.1'
>>> torch._C._get_default_device()
'cpu'
>>> torch.set_default_device('cuda')
>>> torch._C._get_default_device()
'cpu'
>>> torch.set_default_tensor_type(torch.cuda.FloatTensor)
>>> torch._C._get_default_device()
'cuda'
```

### Versions

torch==`2.1.1` and also latest `main`

cc @albanD

### LINKED ISSUE #115917: torch.set_default_device fails in multi-threading
### 🐛 Describe the bug

As the title stated.

`torch.set_default_device` considers multi-threading scenarios (TLS) at the C++ level, but not at the Python level.

**Code**

```Python
from threading import Thread

import torch

def task1():
    torch.set_default_device('cuda')
    print(torch.tensor([1, 2, 3]).device)


if __name__ == '__main__':
    torch.set_default_device('cpu')
    print(torch.tensor([1, 2, 3]).device)

    t1 = Thread(target=task1)
    t1.start()
    t1.join()
```

**Output**:

```Python
cpu
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/dev/lib/python3.9/threading.py", line 980, in _bootstrap_inner
    self.run()
  File "/usr/local/anaconda3/envs/dev/lib/python3.9/threading.py", line 917, in run
    self._target(*self._args, **self._kwargs)
  File "/root/a.py", line 6, in task1
    torch.set_default_device('cuda')
  File "/root/Git.d/pytorch/pytorch/torch/__init__.py", line 663, in set_default_device
    _GLOBAL_DEVICE_CONTEXT.__exit__(None, None, None)
  File "/root/Git.d/pytorch/pytorch/torch/utils/_device.py", line 71, in __exit__
    return super().__exit__(exc_type, exc_val, exc_tb)
  File "/root/Git.d/pytorch/pytorch/torch/overrides.py", line 1904, in __exit__
    _pop_mode()
  File "/root/Git.d/pytorch/pytorch/torch/overrides.py", line 1927, in _pop_mode
    old = _pop_torch_function_stack()
RuntimeError: trying to pop from empty mode stack
```

### Versions

Collecting environment information...
PyTorch version: 2.3.0a0+git1310f0b
Is debug build: True
CUDA used to build PyTorch: 12.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
Clang version: Could not collect
CMake version: version 3.27.0
Libc version: glibc-2.35

Python version: 3.9.16 (main, May 15 2023, 23:46:34)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-73-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.128
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: Tesla P4
GPU 1: Tesla P4

Nvidia driver version: 535.86.10
cuDNN version: Probably one of the following:
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn.so.8.9.4
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.9.4
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.9.4
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.9.4
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.9.4
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.9.4
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.9.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   42 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          16
On-line CPU(s) list:             0-15
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) CPU E5-2697 v4 @ 2.30GHz
CPU family:                      6
Model:                           79
Thread(s) per core:              2
Core(s) per socket:              4
Socket(s):                       2
Stepping:                        1
BogoMIPS:                        4599.99
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       256 KiB (8 instances)
L1i cache:                       256 KiB (8 instances)
L2 cache:                        2 MiB (8 instances)
L3 cache:                        90 MiB (2 instances)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-15
Vulnerability Itlb multihit:     KVM: Mitigation: VMX unsupported
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:          Vulnerable
Vulnerability Mmio stale data:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown

Versions of relevant libraries:
[pip3] botorch==0.8.5
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-coding==1.3.3
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] gpytorch==1.10
[pip3] mypy==1.7.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] onnx==1.14.0
[pip3] onnxruntime-gpu==1.15.1
[pip3] onnxscript==0.1.0.dev20231109
[pip3] optree==0.10.0
[pip3] pytorch-lightning==2.0.6
[pip3] pytorch-sphinx-theme==0.0.24
[pip3] torch==2.3.0a0+git9e84d0f
[pip3] torchmetrics==1.0.1
[pip3] torchmultimodal-nightly==2023.7.31
[pip3] torchrl==0.1.1
[pip3] torchx==0.5.0
[pip3] triton==2.0.0.post1
[conda] botorch                   0.8.5                    pypi_0    pypi
[conda] gpytorch                  1.10                     pypi_0    pypi
[conda] numpy                     1.26.0                   pypi_0    pypi
[conda] optree                    0.10.0                   pypi_0    pypi
[conda] pytorch-lightning         2.0.6                    pypi_0    pypi
[conda] pytorch-sphinx-theme      0.0.24                    dev_0    <develop>
[conda] torch                     2.3.0a0+git9e84d0f           dev_0    <develop>
[conda] torchfix                  0.2.0                    pypi_0    pypi
[conda] torchmetrics              1.0.1                    pypi_0    pypi
[conda] torchmultimodal-nightly   2023.7.31                pypi_0    pypi
[conda] torchrl                   0.1.1                    pypi_0    pypi
[conda] torchx                    0.5.0                    pypi_0    pypi
[conda] triton                    2.0.0.post1              pypi_0    pypi

cc @albanD

### PR DESCRIPTION
**Pull request by fffrog**
1. fix bug of torch.set_default_device in multi-threading
2. add new interface named torch.get_default_device

Fixes #115333
Fixes #115917

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by malfet — CHANGES_REQUESTED** [review]
[Edit] Actually, I see that there was already a discussion about it, so probably it's fine to have both as part of the same PR

File: test/test_utils.py [hunk 1]
```diff
@@ -889,21 +888,43 @@ def f():
...
 
         self.assertEqual(r.device.type, 'meta')
 
+    def test_get_default_device(self):
+        torch.set_default_device('meta')
+        self.assertEqual(torch.get_default_device().type, 'meta')
+        torch.set_default_device(None)
+
+    @unittest.skipIf(
+        not (torch.cuda.is_available() and torch.cuda.device_count() > 1),
```
Comment (malfet) [inline test/test_utils.py:911]: Please use `TEST_MULTIGPU` from `torch.testing._internal.common_cuda`
```suggestion
        not TEST_MULTIGPU,
```

**Review by malfet — APPROVED** [review]

**Review by jansel — APPROVED** [review]

File: test/test_utils.py (hunk 1 above)
Comment (FFFrog) [reply test/test_utils.py]: Done :D

**Pull request closed without merging**